test/functional/api/keymap_spec.lua

```diff
@@ -0,0 +1,243 @@
+        buffer_on_result, option, option_token, map, mode), function()
+      make_new_windows(new_windows)
+      nvim('command', map .. ' <buffer> ' .. option_token .. ' foo bar')
+      local result = nvim('buf_get_keymap', 0, mode)[1][option]
+      eq(buffer_on_result, result)
+    end)
+
+    -- Don't run these for the <buffer> mapping, since it doesn't make sense
+    if option_token ~= '<buffer>' then
+      it(string.format('returns %d for the key "%s" when %s is not used globally with %s (%s)',
+          global_off_result, option, option_token, map, mode), function()
+        make_new_windows(new_windows)
+        nvim('command', map .. ' baz bat')
+        local result = nvim('get_keymap', mode)[1][option]
+        eq(global_off_result, result)
+      end)
+
+      it(string.format('returns %d for the key "%s" when %s is not used for buffers with %s (%s)',
+          buffer_off_result, option, option_token, map, mode), function()
+        make_new_windows(new_windows)
+        nvim('command', map .. ' <buffer> foo bar')
+
+        local result = nvim('buf_get_keymap', 0, mode)[1][option]
+        eq(buffer_off_result, result)
+      end)
+    end
+  end
+
+  -- Standard modes and backwards compatible
+  local mode_list = {'nnoremap', 'nmap', 'imap', 'inoremap', 'cnoremap'}
+  for mode in pairs(mode_list) do
+    global_and_buffer_test(mode_list[mode], 'silent', '<silent>', 1, 1, 0, 0)
+    global_and_buffer_test(mode_list[mode], 'nowait', '<nowait>', 1, 1, 0, 0)
+    global_and_buffer_test(mode_list[mode], 'expr', '<expr>', 1, 1, 0, 0)
+  end
+
+  -- noremap will now be 2 if script was used, which is not backwards compatible
+  global_and_buffer_test('nmap', 'noremap', '<script>', 2, 2, 0, 0)
+  global_and_buffer_test('nnoremap', 'noremap', '<script>', 2, 2, 1, 1)
+
+  -- buffer will return the buffer ID, which is not backwards compatible
+  -- Three of these tests won't run
+  global_and_buffer_test('nnoremap', 'buffer', '<buffer>', nil, 3, nil, nil, 2)
+
+  it('returns script numbers for global maps', function()
+    source([[
+    function! s:maparg_test_function() abort
+      return 'testing'
+    endfunction
+
+    nnoremap fizz :call <SID>maparg_test_function()<CR>
+    ]])
+    eq(1, nvim('get_keymap', 'n')[1]['sid'])
+    eq('testing', nvim('call_function', '<SNR>1_maparg_test_function', {}))
+  end)
+
+  it('returns script numbers for buffer maps', function()
+    source([[
+    function! s:maparg_test_function() abort
+      return 'testing'
+    endfunction
+
+    nnoremap <buffer> fizz :call <SID>maparg_test_function()<CR>
+    ]])
+    eq(1, nvim('buf_get_keymap', 0, 'n')[1]['sid'])
+    eq('testing', nvim('call_function', '<SNR>1_maparg_test_function', {}))
+  end)
+
+  it('works with <F12> and others', function()
+    source([[
+    let g:maparg_test_var = 0
+
+    nnoremap <F12> :let g:maparg_test_var = 1<CR>
+    ]])
+    eq(0, eval('g:maparg_test_var'))
+    source([[
+    call feedkeys("\<F12>")
+    ]])
```

Comment:
There is `[helpers.]feed('<F12>')`, though checking that mapping works is not required. `source` is an overkill unless you do need this: it is working with a temporary file thus is slow and should be omitted when possible.